### PR TITLE
Fixed #30070 -- Fixed content spoofing possiblity in the default 404 page.

### PR DESCRIPTION
**Pull request by tasn**
This issue can lead to phishing, SEO injection and other issues. More
information is available on the OWASP website:
https://www.owasp.org/index.php/Content_Spoofing

How it looks in django with the fix applied (you can use the try the
same url with any django app to see pre-fix):
https://www.etesync.com/account%20was%20not%20found%20on%20this%20server.%20This%20usually%20means%20an%20issue%20with%20your%20account%20that%20may%20lead%20to%20data%20loss!%20Please%20contact%20support%20ASAP%20at%20hacker@mail.com%20to%20fix%20your%20account%20and%20report%20that%20your%

The fix:
Remove the path from the default template so it just includes a message
without the path.

The issue has been initially reported to me about my Django app running at
www.etesync.com by Jerbi Nessim. After investigating the issue I found
that the cause is Django's default 404 handler, and wrote this fix.

https://code.djangoproject.com/ticket/30070

### CODE REVIEW

**Comment by pope1ni** [comment]
Please can you create a ticket and add a test as per the [contributing guidelines](https://docs.djangoproject.com/en/stable/internals/contributing/)?

I agree with using `urlencode`, but I also wonder whether the following might be better:
```patch
-            '<p>The requested URL {{ request_path }} was not found on this server.</p>')
+            '<p>The requested URL was not found on this server: <tt>{{ request_path|urlencode }}</tt></p>')
```
This gets rid of the link and changes the position and styling of the text to make it clearer that any surplus cruft has been added to the URL and is not part of the main text of the error page.

**Comment by tasn** [comment]
Sorry, I thought this falls under the trivial/typo fixes category (even though it has more significant implications), which seemed to not require a ticket under the guidelines. I'll take a look into creating a ticket soon.

My concern with urlencode was about usability. As for most users, a urlencoded path may as well be gibbrish, especially when using non-latin alphabets.
I like the idea of re-arranging the order, it's another variation I was playing with, will change it now.

Thanks for your review and comments.

**Comment by tasn** [comment]
Sorry, I just took your change as-is without running the tests, that now fail. I'll have to fix it all tomorrow or the day after.

**Comment by MarkusH** [comment]
Thank you for your reportand patch, @tasn.

I would prefer to not include the path at all and rephrase the message to something like "The requested resource was not found on this server."

For the future, may I kindly ask you to report potential security issues to `security@djangoproject.com`, according to [the Django Security Policy](https://docs.djangoproject.com/en/2.1/internals/security/#reporting-security-issues). Thank you.

**Comment by apollo13** [comment]
> I would prefer to not include the path at all and rephrase the message to something like "The requested resource was not found on this server."

sounds good to me. Fwiw by default Nginx does not include the URL and Apache includes the requested URL but without an a-tag.

**Comment by timgraham** [comment]
The default template dates to 8bd7b598b6de1be1e3f72f3a1ee62803b1c02010 and `request_path` is included in the template context before that.

I'm also in favor of not including the path in the default template but it could also make sense to apply `urlencode` to `request_path` in the view so that custom 404 templates don't have to do that.

**Comment by MarkusH** [comment]
Good idea, @timgraham.

**Comment by tasn** [comment]
@apollo13, the a tag was just to visually differentiate the link from the rest of the content in a way that an attacker can't control. Used the a tag because it was used on a link, so it made visual sense. It's however not needed if you urlencode the path.

@timgraham, I agree and I'm happy to send another PR, though this is out of scope for this PR, no?
Additionally, this is a breaking change that affects user code. Users could have, for example, already urlencoded the path in their templates (like I have in EteSync). How do you handle such breaking changes?

Anyhow, I just pushed the fixed PR.

**Comment by MarkusH** [comment]
> @timgraham, I agree and I'm happy to send another PR, though this is out of scope for this PR, no?
> Additionally, this is a breaking change that affects user code. Users could have, for example, already urlencoded the path in their templates (like I have in EteSync). How do you handle such breaking changes?

Since this is a security relevant change, we can, if there's no easy way around, break backwards compatibility.

**Comment by tasn** [comment]
I just pushed an additional commit that urlencodes the path in the context, so I think this PR is now ready to be merged.

Please let me know if you have any other comments.

**Comment by timgraham** [comment]
I added release notes and made a few edits that could be reviewed.

File: tests/handlers/tests.py [hunk 1]
```diff
@@ -162,16 +162,17 @@ def test_suspiciousop_in_view_returns_400(self):
 
     def test_invalid_urls(self):
         response = self.client.get('~%A9helloworld')
-        self.assertContains(response, '~%A9helloworld', status_code=404)
+        self.assertEqual(response.status_code, 404)
 
+        # Verify that we don't include the urldecoded path to prevent content spooding
         response = self.client.get('d%aao%aaw%aan%aal%aao%aaa%aad%aa/')
-        self.assertContains(response, 'd%AAo%AAw%AAn%AAl%AAo%AAa%AAd%AA', status_code=404)
+        self.assertNotContains(response, 'd%AAo%AAw%AAn%AAl%AAo%AAa%AAd%AA', status_code=404)
```
[the remark this answers is not in the document]
Comment (timgraham) [reply tests/handlers/tests.py]: Done, this a comment from an older version of the patch.

[the remark this answers is not in the document]
Comment (MarkusH) [reply tests/handlers/tests.py]: Yeah, just realized that as well. Thanks for keeping the comment, GitHub 😀

**Comment by tasn** [comment]
Awesome, looking forward to having it in!

I see the commit message omits all of the information about content spoofing and the example failed url, and it also doesn't point back to this PR. Is this intentional?

**Comment by pope1ni** [comment]
Yes. The commit message refers to the Trac ticket. The Trac ticket refers to this PR. (When it is merged, viewing the commit in GitHub will also show the PR and tags the commit is in...)

**Pull request merged**